Boost: handle post edits and comments and delete cache files

donnchawp opened the pull request:
The Cache module has to delete cache files when posts are edited, comments are made or other actions are taken that modify the website.
This PR will take care of those actions and delete the right files so cache files don't grow stale.

Fixes [#35055](https://github.com/Automattic/jetpack/issues/35055)

## Proposed changes:
* Lots of error_log statements still in the code to aid testing. I'll have to remove them, and get an approval again.
* Hook onto various actions so the cache can be cleared when posts are edited, comments made or the theme is switched. We'll probably need more.
* In is_cacheable() check if the user is viewing a preview, or using the customizer. Not applicable right now because we don't cache for logged in users. Maybe in future?
* In normalize_request_uri(), parse_url may return null or false, so we need to handle those.
* Added the function is_visible_post_type( $post ) that returns true if the post type of that post is visible.
* Add the function delete_front_page_cache( $post ) because the front page should be handled separately. It can be in a number of formats (static page, or page of blog posts..)
* Added functions to handle comment editing and publishing.
* The post status can be either published or private which are visible on the web page, or a number of other statuses which aren't.  The delete_on_post_transition() function should take care of that and will delete the related tags and categories, and the homepage.
* delete_cache_for_post() was added to accept a $post, get a permalink, even if the post was trashed, and then delete the cache files for that post.
* delete_cache_for_post_terms() will delete the tag and category archives for the given post.
* delete_cache_for_url( $url ) deletes the cache files for that url, normalizing and then passing on the URI to the storage class.
* delete_cache() will delete the entire cache for the site.
* In the Utils library, the delete_directory() function gets a new parameter, $recurse. If it's 1 then it will delete any directories found by recursing down through them.
* delete_single_directory() simply calls delete_directory() with $recurse set to false.
* In the storage class, invalidate( $path ) will delete the path, recursively.
* invalidate_single_visitor() will delete the cache files for the given path, and user with GET and COOKIEs.
* invalidate_home_page() deletes the html files found in the root of the cache directory, and any "page" archives.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pc9hqz-2rA-p2

## Does this pull request change what data or activity we track or use?
no

## Testing instructions:
* Create some tags.
* Create some test posts and add them to the tags.
* Populate the cache by visiting these posts, and the homepage too. You must use a private, NOT LOGGED IN, browser window.
* Tail the logs with `jetpack docker tail`.
* Edit the cache files and watch the tailed log. It should show it deleting that post, and the home page. Please verify the cache files were deleted from the directory.
* Change a post from published to draft. Change it from either of those to trashed. Restore it from trash. No cache files should be edited if the post isn't published either before or after.
* Leave a comment on a post. If moderated, you'll see a log entry saying the cache for that particular user was deleted.
* Try again.
* In wp-admin->comments, moderate the comments. Approve and reverse that. Edit the comment. Trash it. Examine the logs.
* Switch theme and note it deletes the entire cache.
* Link to another post on your site while editing a post. The moderated pingback will appear in the comments page. Approve it and note the cached pages being deleted.

donnchawp committed:
WIP in progress on handling edits in case anyone wants to see

donnchawp committed:
changelog

donnchawp commented:
Let's see what happens with https://github.com/Automattic/jetpack/pull/35515 before doing any more on this one.

donnchawp committed:
Merge branch 'trunk' into boost-cache/handling-edits

donnchawp committed:
Major cleanup of edit handling

Can delete single directories now, so the homepage can be deleted.
Also, single posts can be deleted when a comment status changes.
If someone leaves an moderated comment, only their cache file will
be deleted.

donnchawp committed:
Updated docblock

donnchawp committed:
Removed error_log debugging

donnchawp committed:
Add switch_theme support, delete the entire cache

donnchawp committed:
Only delete the post itself when comments are modified

donnchawp committed:
Add delete_cache_for_post_only() to delete the post, not homepage

donnchawp committed:
Use delete_cache_for_post_only() on comment post

Fix checking the comment_approved status

donnchawp committed:
Merge branch 'trunk' into boost-cache/handling-edits

donnchawp committed:
Add invalidate() function to delete a path and subdirectories

donnchawp committed:
Added invalidate() functions and some logging

donnchawp committed:
Add some logging

donnchawp committed:
Added back in deletion functions.

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into boost-cache/handling-edits

donnchawp committed:
Delete the cache for the right URL, the post URL, not comment php file

donnchawp committed:
Add $recurse to logging to make testing easier

donnchawp committed:
Delete /page/ when deleting the homepage.

donnchawp committed:
Handle editing comments

donnchawp committed:
Simplify homepage maintenance as paged archives deleted.

There's not much point in checking if the the edited post is on the
front page if it's on the second page of the archives. So we should
delete the homepage and page archives.

donnchawp committed:
Updates to doc blocks

donnchawp committed:
Add delete_cache function to delete everything.

Used when switching themes

donnchawp committed:
Update doc block

donnchawp committed:
Simplify the post status transition action.

Only delete the cache if the old or new status was published in some
way.

donnchawp committed:
Remove this code as we'll rely on post transition status

dilirity reviewed (commented):
Did a quick skim, haven't tested anything yet.

Do we need all the functions we hook on `init_actions` inside the `Boost_Cache` class?

We could attack this in an upcoming PR, but I was wondering if it makes more sense to have those things somewhere else, as they seem pretty specific.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	 * @return bool - True if the post type is public.
+	 */
+	protected function is_visible_post_type( $post ) {
+		$post_type = is_object( $post ) ? get_post_type_object( $post->post_type ) : null;
```
Instead of `is_object( $post) `, I'd suggest using `is_a( $post, 'WP_Post' )`, as we're specifically looking for properties in there.

donnchawp replied:
Good idea, done!

donnchawp commented:
> Did a quick skim, haven't tested anything yet.
> 
> Do we need all the functions we hook on `init_actions` inside the `Boost_Cache` class?
> 
> We could attack this in an upcoming PR, but I was wondering if it makes more sense to have those things somewhere else, as they seem pretty specific.

@dilirity I was thinking about that too as it does make that file significantly longer. Maybe an "Edit" class? "Boost_Cache_Edit"?

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	 *
+	 * @param int $post_id - The id of the post that was edited, deleted, or published.
+	 */
+	protected function delete_front_page_cache( $post ) {
+		$front_page_id = get_option( 'show_on_front' ); // posts page
+		if ( $front_page_id === 'page' ) {
+			$front_page_id = get_option( 'page_on_front' ); // static page
+			if ( $front_page_id === $post->ID ) {
+				$this->delete_cache_for_post( $post->ID );
+			}
+		} else {
+			$this->storage->invalidate_home_page();
+			error_log( 'delete front page cache' ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
+		}
+	}
```
Shouldn't deleting the front page always going to call `invalidate_home_page`? Maybe I'm missing something, but I'd think that's what it should do. Why do we need to clear the cache for a specific post?

donnchawp replied:
This is for when you're using a static page for the front page. (See Settings->Reading in wp-admin)

dilirity replied:
Isn't the cache file the one for `/`?

thingalon replied:
No; the site won't always be at `/`; it can be in a subdirectory, for example. 

But that then leaves me wondering how `invalidate_home_page` works, as it deletes the root of the cache directory, which may or may not be the site's homepage.

donnchawp replied:
You're both right. While I was sitting in the driver's seat in my car waiting for the roadside assistance yesterday, I tested using a static page for the front page. It was cached as "/", so that function can be simplified.
I'm going to test this PR on my test site in a subdirectory and see how that works and make adjustments.

donnchawp replied:
So. Two issues here:
* Site is hosted at /test/ instead of /
* Site is using a static page for the front page and another page for posts

I'll handle the second one first. The first "front_page_id" check above handled that, but it didn't work at all. It did update the static homepage, but never updated the page used to hold the posts.

I've just rewritten that. The static front page is at "/" (or "/test/", it doesn't matter as the post permalink has the right directory) and the posts page in my test example is at /posts/. Both locations are cleared when a post is updated. There might be a case to be made that the static homepage doesn't need to be updated, but some sites will want it updated.

donnchawp replied:
I just fixed having a site hosted at /test/ rather than / by passing the path from the home_url() to invalidate_home_page().

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	 * Delete the cache for the front page and paged archives.
+	 * This is called when a post is edited, deleted, or published.
+	 *
+	 * @param int $post_id - The id of the post that was edited, deleted, or published.
```
I think this is outdated, since `$post_id` is a `WP_Post` instance.

thingalon replied:
If `post_id` is a `WP_Post` instance, should it be called `post`?

donnchawp replied:
Yes, definitely. I missed that when I updated that function.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+
+		error_log( "delete_on_post_transition: deleting post {$post->ID}" ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
+
+		$this->delete_cache_for_post( $post );
+		$this->delete_cache_for_post_terms( $post );
+		$this->delete_front_page_cache( $post );
```
I think naming can be improved here and they should all follow the same structure.

For example, `delete_cache_THING`.

Does this make sense?

donnchawp replied:
Yep, renamed it to `delete_cache_for_front_page`.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+
+		error_log( "delete_on_post_transition: $new_status, $old_status, {$post->ID}" ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
+
+		// Don't delete the cache for posts that weren't published and aren't published now
+		if ( ! $this->is_published( $new_status ) && ! $this->is_published( $old_status ) ) {
```
I get a headache just trying to understand this. It's not your fault, it's just these post transition hooks are such a messy thing.

The thing is now this, but is going to be that, but what was it before? 🤒

donnchawp replied:
The comparisons used to be a lot more complicated until I realised that I only needed to make sure that the post in question was never published, in the past, or if it was about to be.

In other words:
If a post was published and now isn't OR wasn't published and now is, then clear the cache for that post.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -51,6 +51,19 @@ public function __construct( $storage = null ) {
+		$this->init_actions();
+	}
+
+	/*
```
I've noticed that you haven't been adding proper docblock comments. They all should start with `/**`. That way, the editor properly picks it up and colors it, showing the variables and return statements in a neat color.

donnchawp replied:
oh, thanks. I never noticed that.

donnchawp replied:
I made the same error in Page_Cache and the Utils files but I won't fix them in this PR, until I need to modify them for something else that is and I remember seeing a comment about a function that should go in Utils ...

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
 		return $is_backend;
 	}
+
+	/*
```
As mentioned above, something to be mindful of. Add a second star here and see how the comment changes colors.

donnchawp replied:
Got it, and fixed that. Thanks!

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -157,7 +175,7 @@ public function is_cacheable() {
 		// get path from request uri
 		$request_uri = parse_url( $request_uri, PHP_URL_PATH ); // phpcs:ignore WordPress.WP.AlternativeFunctions.parse_url_parse_url
-		if ( $request_uri === '' ) {
+		if ( $request_uri === '' || $request_uri === false || $request_uri === null ) {
```
Lol, this function can basically return anything 😅

thingalon replied:
Is this different to `is_empty()`?

donnchawp replied:
Yeah, basically empty(). I'll change it to that. It was annoying when I saw that parse_url returns so many false values.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	 * @param WP_Comment $comment - The comment that transitioned.
+	 */
+	public function delete_on_comment_transition( $new_status, $old_status, $comment ) {
+		if ( $new_status === $old_status ) {
```
If a comment was `pending`, but was marked as `spam`, this would also reset the cache. Should it? I can't recall what happened when you added a comment to a page. I guess it shows up only for the user, but with a "pending approval" text.

donnchawp replied:
You're right. I forgot about spam. There's also ['hold', 'approve', or 'trash'](https://developer.wordpress.org/reference/functions/wp_set_comment_status/). I'll have to add checks like in the post transition function.

donnchawp replied:
Fixed that by checking if comment was not approved before and after, and returning if it was.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	 * @param bool $all - If false, only delete the cache file for the post, not the front page cache.
+	 */
+	public function delete_cache_for_post( $post ) {
+		static $already_deleted = -1;
+		if ( $already_deleted === $post->ID ) {
+			return;
+		}
```
Hehe, nice thinking!

thingalon replied:
If we can end up invalidating the same post multiple times in a single PHP execution, could we also end up deleting multiple posts, multiple times: Is it worth storing an array of post_ids that have been invalidated this PHP execution to prevent re-deleting?

I don't know the answer to this question, btw - so if the answer is "no" that's ok.

donnchawp replied:
I copied this from WPSC where there are more actions fired for deleting cache files, but since there are a few comment actions hooked, it's still needed.
I think it's ok to have just one. When I logged the deletes I didn't see the same post being deleted twice with this in place.

pyronaur replied:
I don't know how this works, but it seems to me that doing this would break:

```
delete_cache_for_post( get_post(1) );
delete_cache_for_post( get_post(2) );
delete_cache_for_post( get_post(1) );
delete_cache_for_post( get_post(2) );
```

Even if the code works now, I would use an array of already deleted posts.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	 * @return bool
+	 */
+	private function is_published( $status ) {
+		return $status === 'publish' || $status === 'private';
```
Why does this check for `private` as well?

thingalon replied:
`private` means it's published for administrators on the site. While we currently do not cache any logged in content, we might in the future.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	}
+
+	/*
+	 * Delete the cache for the given url. Recurse through sub directories if $recurse is true.
```
This is outdated. I don't see `$recurse`.

donnchawp replied:
Thanks, fixed that.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache_Utils.php:
```diff
@@ -35,24 +35,26 @@ public static function is_feed() {
 		// make sure that $dir is a directory inside WP_CONTENT . '/boost-cache/';
 		if ( self::is_boost_cache_directory( $dir ) === false ) {
-			return new \WP_Error( 'Invalid directory' );
+			return new \WP_Error( 'Invalid directory ' . $dir );
```
We should make these translatable.

donnchawp replied:
Good point. There are others that need translating too.

dilirity commented on projects/plugins/boost/app/modules/cache/storage/File_Storage.php:
```diff
@@ -102,4 +102,57 @@ private function sanitize_path( $path ) {
+		return false;
+	}
+
+	/**
```
This one should be colored in the editor!

donnchawp replied:
Fixed earlier.

dilirity reviewed (dismissed):
Nice work on this! I haven't run any tests as I'm beat. I'll test it tomorrow.

I've left a few notes.

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	 * @param WP_Post $post - The post to check.
+	 * @return bool - True if the post type is public.
+	 */
+	protected function is_visible_post_type( $post ) {
```
Belongs in `Boost_Cache_Utils`.

donnchawp replied:
Good idea, done.

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+			}
+		} else {
+			$this->storage->invalidate_home_page();
+			error_log( 'delete front page cache' ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
```
I'm ok for the `error_log` statements to remain in here, pending being turned into proper logging statements as Adnan creates the logging system.

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	}
+
+	/*
+	 * Delete the cache for the front page and paged archives.
+	 * This is called when a post is edited, deleted, or published.
```
```suggestion
	 * Delete the cache for the front page if it will be affected by changes to the specified post.
	 * This is called when a post is edited, deleted, or published.
```

donnchawp replied:
I decided to remove the code that checked if the post was on the front page because invalidate_home_page() now deletes the paged archives now, so the comment is correct.

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	 * Deletes cache files for the given post.
+	 *
+	 * @param WP_Post $post - The post to delete the cache file for.
+	 * @param bool $all - If false, only delete the cache file for the post, not the front page cache.
```
There is no `$all` param.

donnchawp replied:
Good catch, thanks!

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	 * @param string $status - The status of the post.
+	 * @return bool
+	 */
+	private function is_published( $status ) {
```
Should this go in `Boost_Cache_Utils`?

donnchawp replied:
It probably doesn't because the only code that checks the status like this will be the post transition function.

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -254,4 +272,208 @@ public function is_backend() {
+	public function delete_on_comment_post( $comment_id, $comment_approved, $commentdata ) {
+		$post = get_post( $commentdata['comment_post_ID'] );
+
+		/*
+		 * If a comment is not approved, we only need to delete the cache for
+		 * this post for this visitor so the unmoderated comment is shown to them.
+		 */
+		if ( $comment_approved !== 1 ) {
+			$this->storage->invalidate_single_visitor( $this->normalize_request_uri( get_permalink( $post->ID ) ), $this->request_parameters );
+			return;
+		}
```
Nice; it's the complex cases like this I would never think of that I really appreciate your experience for :)

donnchawp replied:
Thanks :)

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -51,6 +51,19 @@ public function __construct( $storage = null ) {
+	 * Initialize the actions for the cache.
+	 */
+	protected function init_actions() {
+		add_action( 'transition_post_status', array( $this, 'delete_on_post_transition' ), 10, 3 );
+		add_action( 'transition_comment_status', array( $this, 'delete_on_comment_transition' ), 10, 3 );
+		add_action( 'comment_post', array( $this, 'delete_on_comment_post' ), 10, 3 );
+		add_action( 'edit_comment', array( $this, 'delete_on_comment_edit' ), 10, 2 );
+		add_action( 'switch_theme', array( $this, 'delete_cache' ) );
```
Do we need `save_post` as well?

donnchawp replied:
I don't think we'll need that. The `transition_post_status` action fires when a post is saved, even if the [status hasn't changed](https://developer.wordpress.org/reference/hooks/transition_post_status/#comment-2977).

thingalon reviewed (changes requested):
Oh, man. Working out what to delete for each kind of change is complicated in WordPress.

donnchawp committed:
Check that $post is a WP_Post explicitely.

donnchawp committed:
Fixed docblock on delete_front_page_cache()

donnchawp committed:
Rename for consistency with other similar functions

donnchawp committed:
Fix up docblocks in these two files

donnchawp committed:
More docblock fixes

donnchawp committed:
All these are basically empty()

donnchawp committed:
Remove outdated comment about $recurse

donnchawp committed:
These warnings should be translatable

donnchawp committed:
Fix docblocks

donnchawp committed:
Move is_visible_post_type() to Utils class

donnchawp committed:
Remove unused param from comment

donnchawp committed:
Fix docblock param

donnchawp committed:
Don't delete on comments unless they're approved.

Delete only: 1 unapproved -> approved
             2 approved -> unapproved/trash/spam
Don't delete when unapproved -> trash or spam

Commit:
Merge branch 'trunk' into boost-cache/handling-edits

thingalon commented on projects/plugins/boost/app/modules/cache/storage/File_Storage.php:
```diff
@@ -102,4 +102,57 @@ private function sanitize_path( $path ) {
+	/**
+	 * Delete the cached files for the home page, and any paged archives.
+	 */
+	public function invalidate_home_page() {
+		$directory = $this->root_path;
+		$dir       = Boost_Cache_Utils::sanitize_file_path( $directory );
+
+		if ( Boost_Cache_Utils::is_boost_cache_directory( $dir ) ) {
+			if ( is_dir( $dir . '/page' ) ) {
+				Boost_Cache_Utils::delete_directory( $dir . '/page' );
+			}
+			return Boost_Cache_Utils::delete_single_directory( $dir );
+		}
+
+		return false;
+	}
```
If WordPress is installed in a subdir, or is otherwise not at the `/` URL of their site, then won't this completely fail to clear the cache of their homepage?

Also what is the check for `/page` for? I think that needs a comment explaining what it means.

donnchawp replied:
You're right about a subdirectory. I'm going to test that. /page/ is for paged archives. Here's [page 426](https://inphotos.org/page/426/) of my photoblog. :)

donnchawp replied:
I just fixed that.

thingalon reviewed (dismissed):
Ok, I'm approving this because:
- I think the only issue remaining that I can see is that invalidate_home_page can't know that its home path is `/`
- I will be offline today while @dilirity and @donnchawp will be online to review and make sure this is good to go :) 

One note I will add: please be careful on how this merges with the hostnames update - https://github.com/Automattic/jetpack/pull/35647

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into boost-cache/handling-edits

donnchawp commented:
https://github.com/Automattic/jetpack/pull/35647 merged cleanly, thankfully.

donnchawp committed:
Delete static page/posts page properly

donnchawp committed:
Add home_url() path to invalidate_home_page

This will let it work on sites that are not running off "/"

pyronaur commented on projects/plugins/boost/app/modules/cache/storage/Storage.php:
```diff
@@ -9,4 +9,7 @@ interface Storage {
 	public function write( $request_uri, $parameters, $data );
 	public function read( $request_uri, $parameters );
+	public function invalidate( $request_uri );
+	public function invalidate_single_visitor( $request_uri, $parameters );
+	public function invalidate_home_page( $dir );
```
Because this is an interface, I would caution against requiring the Storage to know about specifics of what it's storing- "single visitor" or "home_page", that should probably be done at the implementation level.

Looking at the surrounding code, It looks to me like the `invalidate` interface should be `invalidate( $request_uri, $parameters)` and the rest should be handled by the implementation instead.

But I haven't followed along the developments here and I'm on SR right now, so just adding this as a comment here.

thingalon replied:
I understand where you're coming from, and I agree that we should think about what each class is responsible for.

But in this specific instance, I don't think you're correct. The Storage class is specific to caching, and exists to understand how to store data indexed by its metadata, in a format where it can have segments of its cache wiped based on their relationships.

I actually do think it's important for the Storage interface here to understand the semantics of the data it's storing, because the decisions it makes about how to store it directly stems from what the data is, and how different bits of the data relate to each other.

pyronaur reviewed (approved):
Approving this to allow merging, but I haven't reviewed this (I'm on SR) 🙈
